rackcity/views/power_views.py

```diff
@@ -596,30 +558,75 @@ def make_bcman_request(chassis, blade, power_command):
         )
     try:
         blade_slot = int(data["blade_slot"])
         chassis_id = int(data["chassis_id"])
     except ValueError:
-        raise ChassisPowerManagementException(
+        raise PowerManagementException(
             "Parameters 'chassis_id' and 'blade_slot' must be of type int."
         )
     try:
         chassis = Asset.objects.get(id=chassis_id)
     except ObjectDoesNotExist:
-        raise ChassisPowerManagementException(
-            "Chassis" + GenericFailure.DOES_NOT_EXIST.value
+        raise PowerManagementException("Chassis" + GenericFailure.DOES_NOT_EXIST.value)
+    if not user_has_power_permission(request.user, asset=chassis):
+        raise UserPowerPermissionException(
+            "User "
+            + request.user.username
+            + " does not have power permission and does not own asset with id="
+            + str(data["id"])
         )
-    if (not chassis.model.is_blade_chassis) or (chassis.model.vendor != "BMI"):
-        raise ChassisPowerManagementException(
-            "Power is only network controllable for blade chassis of vendor 'BMI'."
+    if not is_asset_power_controllable_by_bcman(chassis):
+        raise PowerManagementException(
+            "Power is only network controllable for blade chassis of vendor 'BMI' with hostname not in storage."
         )
     if (blade_slot < 1) or (blade_slot > 14):
-        raise ChassisPowerManagementException(
+        raise PowerManagementException(
             "Blade slot " + str(blade_slot) + " does not exist on chassis."
         )
-    return chassis.hostname, blade_slot
+    return chassis, blade_slot
+
+
+def get_pdu_power_request_parameters(request):
+    data = JSONParser().parse(request)
+    if "id" not in data:
+        raise PowerManagementException("Must specify 'id' on pdu power request.")
+    try:
+        asset = Asset.objects.get(id=data["id"])
+    except ObjectDoesNotExist:
+        raise PowerManagementException("Asset" + GenericFailure.DOES_NOT_EXIST.value)
+    if not user_has_power_permission(request.user, asset=asset):
+        raise UserPowerPermissionException(
+            "User "
+            + request.user.username
+            + " does not have power permission and does not own asset with id="
+            + str(data["id"])
+        )
+    if not is_asset_power_controllable_by_pdu(asset):
+        raise PowerManagementException(
+            "Power is only network controllable for assets in datacenter 'RTP1'."
+        )
+    return asset
+
+
+def is_asset_power_controllable_by_bcman(asset):
+    return (
+        asset.model.is_blade_chassis()
+        and not asset.is_in_offline_storage()
+        and asset.model.vendor == "BMI"
+        and asset.hostname is not None
+    )
+
+
+def is_asset_power_controllable_by_pdu(asset):
+    return (
+        asset.model.is_rackmount()
+        and not asset.is_in_offline_storage()
+        and asset.rack.is_network_controlled
+    )
```

Comment:
I think these are good candidates for class functions on the model but I'm fine with them here as well

Reply:
ok yeah I TRIED THAT because I thought so too but it wouldn't let me call `self.model.is_blade_chassis()` or any other methods on the model field, it was expecting a `self` parameter to be passed in, which ain't how class methods work. didn't investigate why but maybe it's a model thing?

Reply:
Wow great minds